src/JMS/Serializer/NormalizerInterface.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * Copyright 2013 Johannes M. Schmitt <schmittjoh@gmail.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace JMS\Serializer;
+
+/**
+ * Interface NormalizerInterface
+ * @author Daniel Bojdo <daniel@bojdo.eu>
+ */
+interface NormalizerInterface
+{
+    /**
+     * Converts objects to an array structure.
+     *
+     * This is useful when the data needs to be passed on to other methods which expect array data.
+     *
+     * @param mixed $data anything that converts to an array, typically an object or an array of objects
+     * @param SerializationContext $context
+     *
+     * @return array
+     */
+    public function toArray($data, SerializationContext $context = null);
+
+    /**
+     * Restores objects from an array structure.
+     *
+     * @param array $data
+     * @param string $type
+     * @param DeserializationContext $context
```

Comment:
True, but sometimes I remember the Java logic that an object can be null, and then avoid the ugly `|null` for object, which is anyway obvious when reading the method declaration in that case